src/Stratis.Bitcoin.Features.PoA/FederationManager.cs

```diff
@@ -20,45 +22,91 @@ public class FederationManager
 
         private readonly ILogger logger;
 
-        public FederationManager(NodeSettings nodeSettings, Network network, ILoggerFactory loggerFactory)
+        private readonly IKeyValueRepository keyValueRepo;
+
+        /// <summary>Key for accessing list of public keys that represent federation members from <see cref="IKeyValueRepository"/>.</summary>
+        private const string federationMembersDbKey = "fedmemberskeys";
+
+        private List<PubKey> federationMembers;
+
+        public FederationManager(NodeSettings nodeSettings, Network network, ILoggerFactory loggerFactory, IKeyValueRepository keyValueRepo)
         {
             this.settings = Guard.NotNull(nodeSettings, nameof(nodeSettings));
             this.network = Guard.NotNull(network as PoANetwork, nameof(network));
+            this.keyValueRepo = Guard.NotNull(keyValueRepo, nameof(keyValueRepo));
 
             this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
         }
 
+        // TODO
+        /*
+         * Subscribe to VotingManager and track when new member is added or deleted. Update keys and then persist.
+         */
+
         public void Initialize()
         {
-            this.LoadKey();
+            // Load federation from the db.
+            this.federationMembers = this.LoadFederationKeys();
 
-            if (this.FederationMemberKey != null)
+            if (this.federationMembers == null)
             {
-                // Loaded key has to be a key for current federation.
-                if (!this.network.ConsensusOptions.FederationPublicKeys.Contains(this.FederationMemberKey.PubKey))
-                {
-                    string message = "Key provided is not registered on the network!";
+                this.logger.LogDebug("Federation members are not stored in the db. Loading genesis federation members.");
```

Comment:
You should probably check here that consensus is on the genesis block, otherwise if the chain is ahead you lose any added or removed members.

Reply:
Not sure if needed. The only case when this is executed is on the initial node start.